fhir-client/src/main/java/uk/ac/ox/ndph/mts/fhir_client/FhirContextWrapper.java

```diff
@@ -0,0 +1,152 @@
+     * @param allPages Go through all pages to return all matching results.
+     * @param <T> The type of resource to search for.
+     * @return A list of resources found.
+     */
+    public <T extends IBaseResource> List<T> search(Class<T> resourceClass,
+                                                    boolean allPages) {
+        return search(resourceClass, null, allPages);
+    }
+
+    /**
+     * Search for resources of specific type without any conditions.
+     *
+     * @param resourceClass The type of resource to search for.
+     * @param criterion The conditions to apply with the search query.
+     * @param allPages Go through all pages to return all matching results.
+     * @param <T> The type of resource to search for.
+     * @return A list of resources found.
+     */
+    public <T extends IBaseResource> List<T> search(Class<T> resourceClass,
+                                                    ICriterion<?> criterion,
+                                                    boolean allPages) {
+        IGenericClient client = getClient();
+
+        IQuery<Bundle> query = client
+                .search()
+                .forResource(resourceClass)
+                .returnBundle(Bundle.class);
+
+        if (criterion != null) {
+            query = query.where(criterion);
+        }
+
+        Bundle resultsBundle = query.execute();
+        List<T> searchResults = BundleUtil.toListOfResourcesOfType(fhirContext, resultsBundle, resourceClass);
+
+        if (allPages) {
+            // loop on next pages
+            while (resultsBundle.getLink().size() > 1) {
+                resultsBundle = client.loadPage()
+                        .next(resultsBundle)
+                        .execute();
+
+                searchResults.addAll(BundleUtil.toListOfResourcesOfType(fhirContext, resultsBundle, resourceClass));
+            }
+        }
+
+        return searchResults;
+    }
+
+    /**
+     * Reads a specific resource by its Id
+     *
+     * @param resourceClass The type of resource to search for.
+     * @param id The id of the resource required.
+     * @param <T> The type of resource to search for.
+     * @return The resource specified.
+     * @throws ResourceNotFoundException When a resource with the type & id isn't found.
+     */
+    public <T extends IBaseResource> T readById(final Class<T> resourceClass,
+                                                final String id) throws ResourceNotFoundException {
+        var baseResource = getClient()
+                .read()
+                .resource(resourceClass)
+                .withId(requireNonBlank(id, "id"))
+                .execute();
+
+        return resourceClass.cast(baseResource);
+    }
+
+    private IGenericClient getClient() {
+        if (this.client == null) {
+            client = fhirContext.newRestfulGenericClient(requireNonBlank(fhirUri, "fhirUri"));
+        }
+        return client;
+    }
```

Comment:
Lazy init of member `client`, means this class is not threadsafe - should be documented at least. Not sure of when instances of FhirContextWrapper are created because that will be done by Spring. Is the IGenericClient threadsafe anyway (docs are unclear)? Is the FhirContext threadsafe (docs say yes and should be reused)?

Reply:
Current implementations we have create their own IGenericClient on each method. We can you with that option again...
The doc is here: https://hapifhir.io/hapi-fhir/docs/client/generic_client.html
It says both the context and the client are thread safe and can be reused.

Regarding how spring is doing this - I believe this is a singleton, so once created it remains a single instance for the during of the application.

Reply:
Another option would be to put this in the constructor...

Reply:
Even if the client is threadsafe, the lazy-initialization is not (because it will presumably be done in a request thread) - so creating the client should be moved to an initializer or the constructor (so will be done only once, and threadsafely, since an object can only be created in one thread). The docs say that creation of the client is cheap, so there is no gain in delaying initialization to first use.

Reply:
@tamirkamara realise this is a work in progress, just a reminder, we'll need to cope with timeouts, when a IGenericClient fails to be initialised.

For the FHIR spike, I looked at 

https://github.com/NDPH-ARTS/fhir-service/blob/main/src/main/java/uk/ac/ox/ctsu/arts/fhirservice/helpers/FHIRClientHelper.java

Reply:
@alexbaxter I've update the constructor to also include the client init. I removed the other overload since that was used for unit tests and we don't plan/can have a real unit text here anyway.

Reply:
@shajidissa what do you mean in cope? Do you mean retry?